### PR TITLE
Fixed #31732 -- Cached callables signatures in django.utils.inspect methods.

### PR DESCRIPTION
**Pull request by timpark0807**
### Overview
**Ticket:** https://code.djangoproject.com/ticket/31732

Created a wrapper function to perform signature inspect lookup and added `lru_cache` decorator to eliminate repeated function calls. Replaced direct signature inspect calls with the cached wrapper function. 

### Testcases 
I am not sure how to add test cases for this functionality. As a manual check, I ran the existing set of test cases with print statements in the wrapper function, with and without the lru_cache. All duplicated calls are eliminated when using the LRU cache.

**Without LRU Cache**
Without LRU Cache

**With LRU Cache** 
With LRU Cache

### CODE REVIEW

File: django/utils/inspect.py [hunk 1]
```diff
@@ -1,8 +1,14 @@
+import functools
 import inspect
 
 
+@functools.lru_cache(None)
+def get_signature(func):
+    return inspect.signature(func)
```
Comment (charettes) [inline django/utils/inspect.py:7]: I initially thought it would be an issue for `boundmethod` instances but it looks like they hash to the same value even if they are bound to different instances

```python
class Foo:
    def __init__(self):
        pass

>>> Foo().__init__ == Foo().__init__
False
>>> hash(Foo().__init__) == hash(Foo().__init__)
True
``` 

In all cases I don't see a compelling reason for setting `maxsize=None` and allowing the cache to grow unbounded.

Comment (timpark0807) [reply django/utils/inspect.py:7]: Thanks for pointing that out. Setting `maxsize=None` was out of pure habit since I normally use lru_cache to memoize my top-down dynamic programming functions.  

How does setting `max_size` to the default value of 128 sound? Thoughts?

File: django/utils/inspect.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
+import functools
 import inspect
 
 
+@functools.lru_cache()
+def get_signature(func):
```
Comment (felixxm) [inline django/utils/inspect.py]: This is an internal wrapper so I changed this to `_get_signature()`.

File: django/utils/inspect.py [hunk 3]
```diff
@@ -1,8 +1,14 @@
+import functools
 import inspect
 
 
+@functools.lru_cache()
```
Comment (felixxm) [inline django/utils/inspect.py]: This wrapper is used in several methods so I've increased `max_size` a bit.

**Comment by Mogost** [comment]
@charettes @timpark0807 
Could you take a look at [this](https://github.com/charettes/django-polymodels/issues/55)?

It's hard for me to tell if it's a django or django-polymodels issue, but it's definitely related to this PR.
This happens because you cannot cache the defaultdict
https://github.com/charettes/django-polymodels/blob/5f40923fd255598131c0f1e48ef6a1f689dd016a/polymodels/models.py#L47

**Comment by charettes** [comment]
@Mogost lets keep the discussion on https://github.com/charettes/django-polymodels/issues/55 as a lot of things break in Django 3.2 for this application. I plan on having a look in the next days but it doesn't look like an issue with Django so far.

**Comment by charettes** [comment]
FWIW this isn't an issue on Python 3.8 but is one on Python <= 3.7 where unhashable type methods are unhashable which prevents their usage with `lru_cache`

```
$ python
Python 3.7.5
Type "help", "copyright", "credits" or "license" for more information.
>>> hash({}.get)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

```
$ python
Python 3.8.5
Type "help", "copyright", "credits" or "license" for more information.
>>> hash({}.get)
16550043
```

So there might be some things to address on the Django side as well since it makes any usage of unhashable types unusable for method inspection on Python < 3.8.

**Comment by charettes** [comment]
@felixxm actually the problem is even worst than I thought, the cache won't be used at all when used on bound methods which is where most the calls originate from

```python
import functools
import inspect

class Foo:
	def __init__(self):
   		pass

@functools.lru_cache(maxsize=512)
def _get_signature(func):
    print('cache miss')
    return inspect.signature(func)

# The following show have been used as a regression test
assert _get_signature(Foo().__init__) is _get_signature(Foo().__init__)
```

**Comment by felixxm** [comment]
@charettes Thanks for details :+1:  I think it's also an argument to "wontfix" ticket :disappointed: 

@adamchainz Can I ask for your opinion?

**Comment by charettes** [comment]
I guess the existing logic could be adjusted to special case bound method by using `.__func__` for introspection instead and returning `parameters` since it looks like the only use of it.

```python
@functools.lru_cache(maxsize=512)
def _get_func_parameters(func, remove_self):
    if remove_self:
        parameters = dict(_get_func_parameters(func, False))
        parameters.pop('self')
        return parameters
    return inspect.signature(func).parameters

def _get_parameters(meth_or_func):
    is_method = inspect.ismethod(meth_or_func)
    func = func.__func__ if is_method else meth_or_func
    return _get_func_parameters(func, remove_self=is_method)
```

**Comment by felixxm** [comment]
I'm not sure it's worth additional complexity :thinking:

**Comment by adamchainz** [comment]
I think it's probably worth it, the performance improvement is pretty significant.

Simon's suggestion looks good, with one caveat: `self` is not a special name, just a convention (one can make a method `def frobnicate(this, self)`...), so we'd want to instead ignore the *first* parameter. It looks like `get_func_full_args` already makes that same mistake.

**Comment by adamchainz** [comment]
I made a PR: #14153.

Admittedly the performance problem doesn't affect any hot path right now since use of `django.utils.inspect` is limited. But I can imagine future deprecations might need to inspect function parameters, in which case I'd like to have the caching versions here and ready to go to prevent performance regressions.

**Comment by Sunkyue-Kim** [comment]
Here is the grammatically corrected and clearer version of the provided text:

---

Applying a cache to the function parameter results in unexpected behavior with `weakref`.

```python
import weakref
from django.utils.inspect import func_accepts_kwargs

def a():
    pass

ref_a = weakref.ref(a)

def a():
    pass

ref_a2 = weakref.ref(a)
func_accepts_kwargs(a)

def a():
    pass

ref_a3 = weakref.ref(a)

print(ref_a)
print(ref_a2)
print(ref_a3)
```

Result:
```
<weakref at 0x000002460DD2F1A0; dead>
<weakref at 0x000002460DD2F2E0; to 'function' at 0x000002460DD14A40 (a)>
<weakref at 0x000002460DD2F240; to 'function' at 0x000002460DD147C0 (a)>
```

The first `weakref` becomes dead, but the second `weakref` remains alive even though the function definition is shadowed.

When redefining a function with the same name, the original function is shadowed. If no references to the original function remain, the original definition is removed, and any `weakref` referencing the original function becomes dead. However, when using that function as a parameter in a context that internally uses `_get_func_parameters`, the function is used as a cache key, resulting in unexpected behavior.

A real-world example related to this issue is signal registration.

Consider the following code in `dispatch/dispatcher.py`:

```python
class Signal:
    ...

    def connect(self, receiver, sender=None, weak=True, dispatch_uid=None):
        from django.conf import settings

        # If DEBUG is on, check that we got a good receiver
        if settings.configured and settings.DEBUG:
            if not callable(receiver):
                raise TypeError("Signal receivers must be callable.")
            # Check for **kwargs
            if not func_accepts_kwargs(receiver):
                raise ValueError(
                    "Signal receivers must accept keyword arguments (**kwargs)."
                )
```

When `DEBUG = True`, the above code block is executed, and the receiver is stored as a cache key. However, when `DEBUG = False`, the code block is not executed, and the receiver is not stored elsewhere.

In this situation:

```python
@receiver(post_save, sender=SomeModel)
def func(sender, instance, **kwargs):
    pass

@receiver(post_save, sender=SomeModel)
def func(sender, instance, **kwargs):
    pass
```

Defining a signal with the same function name results in different behavior depending on the value of `settings.DEBUG`. When `DEBUG = False`, only the last redefinition is executed. However, when `DEBUG = True`, all definitions are executed as signals since each reference to the function definition is used as a cache key, and the internal `weakref` cannot be cleared.

In conclusion, the introduced cache can result in unexpected behavior. Therefore, I believe it should be reverted.

Based on the current state of the latest code, it seems sufficient to simply remove the cache decorator.

**Comment by charettes** [comment]
This relates to https://github.com/python/cpython/issues/88476

I guess a potential solution would be

```diff
diff --git a/django/utils/inspect.py b/django/utils/inspect.py
index 4e065f0347..e194929a80 100644
--- a/django/utils/inspect.py
+++ b/django/utils/inspect.py
@@ -1,15 +1,21 @@
 import functools
 import inspect
+import weakref


 @functools.lru_cache(maxsize=512)
-def _get_func_parameters(func, remove_first):
+def _get_funcweakref_parameters(func_weakref, remove_first):
+    func = func_weakref()
     parameters = tuple(inspect.signature(func).parameters.values())
     if remove_first:
         parameters = parameters[1:]
     return parameters


+def _get_func_parameters(func, remove_first):
+    return _get_funcweakref_parameters(weakref.ref(func), remove_first)
+
+
 def _get_callable_parameters(meth_or_func):
     is_method = inspect.ismethod(meth_or_func)
     func = meth_or_func.__func__ if is_method else meth_or_func
```
